tests/templates/test_subroutines/test_commuting_evolution.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests for the CommutingEvolution template.
+"""
+import pytest
+from pennylane import numpy as np
+import pennylane as qml
+
+
+def test_adjoint():
+    """Tests the CommutingEvolution.adjoint method provides the correct adjoint operation."""
+
+    n_wires = 2
+    dev1 = qml.device("default.qubit", wires=n_wires)
+    dev2 = qml.device("default.qubit", wires=n_wires)
+
+    obs = [qml.PauliX(0) @ qml.PauliY(1), qml.PauliY(0) @ qml.PauliX(1)]
+    coeffs = [1, -1]
+    hamiltonian = qml.Hamiltonian(coeffs, obs)
+    frequencies = [2]
+
+    @qml.qnode(dev1)
+    def adjoint_evolution_circuit(time):
+        for i in range(n_wires):
+            qml.Hadamard(i)
+        # qml.CommutingEvolution(hamiltonian, time, frequencies).adjoint
+        qml.adjoint(qml.CommutingEvolution)(hamiltonian, time, frequencies)
+        return qml.expval(qml.PauliZ(1))
+
+    @qml.qnode(dev2)
+    def evolution_circuit(time):
+        for i in range(n_wires):
+            qml.Hadamard(i)
+        qml.CommutingEvolution(hamiltonian, time, frequencies)
+        return qml.expval(qml.PauliZ(1))
+
+    evolution_circuit(0.13)
+    adjoint_evolution_circuit(-0.13)
+
+    assert all(np.isclose(dev1.state, dev2.state))
+
+
+class TestInputs:
+    """Tests for input validation of `CommutingEvolution`."""
+
+    def test_invalid_hamiltonian(self):
+        """Tests TypeError is raised if `hamiltonian` is not type `qml.Hamiltonian`."""
+
+        invalid_operator = qml.PauliX(0)
+
+        assert pytest.raises(TypeError, qml.CommutingEvolution, invalid_operator, 1)
+
+
+class TestGradients:
+    """Tests that correct gradients are obtained for `CommutingEvolution` when frequencies
+    are specified."""
+
+    def test_two_term_case(self):
+        """Tests the paramer shift rules for `CommutingEvolution` equal the
+        finite difference result for a two term shift rule case."""

```

Comment:
it could be nice here to use [`mocker`](https://github.com/pytest-dev/pytest-mock/) to ensure that the general gradient rule logic is being used :)

E.g., to check that `gen_shift` is being used, something like:

```python
def test_two_term_case(self, mocker):
    spy_gen_shift = mocker.spy(qml.gradients, "gen_shift_rule")
    spy_recipe = mocker.spy(qml.gradients.parameter_shift, "_process_gradient_recipe")
    ...
    grads_param_shift = qml.gradients.param_shift(circuit)(x)

    spy_gen_shift.assert_called()
    shifts, mult, coeffs = spy_recipe.spy_return
    assert shifts == ...
```

Reply:
note that mocker can be very temperamental 🙁 you need to mock the same path that is being imported by the function being spied on